### ISSUE TITLE
[Discussion] How to improve the scoring system?

### ISSUE THREAD

**Issue by pil0u** [pil0u/lewagon-aoc#66]
How it works currently is explained here: https://aoc.lewagon.community/about

What works great:
- Any participant earns points (individually) whenever they solve a puzzle

We identified different issues:
- By nature, groups (batches and cities) have different sizes
- Batches are mostly very small
- All scores are based on the time-to-solve, thus time-zone specific
- In spite of the effort to keep it simple, it requires a 1-page explanation in the [Q&A section](https://aoc.lewagon.community/about)

## Solution 1 - Change batches to groups of fixed size
This requires great changes and raises a number of questions.
- What size for each group?
- Can people come and go from a group to another? Should we lock them at a certain point? When?
- How would it work? A group leader/creator that invites people? Pre-generated groups that people fill?

## Solution 2 - Distinguish hardcore competitors vs casual participants
Hardcore competitor: don't care about getting up early during 25 days, they solve puzzles as soon as they're out
Casual participant: want to try to solve the puzzles through their spare time with no intention to hustle

To enter the hardcore competition and appear on this specific leaderboard, individuals would have to check a box on their profile setting for example (disabled by default).

For casual participants, instead of a _relative solve time_ ranking, we grant fixed points based on time frames:
- If you solve a puzzle (= a _part_) within 24h, you get `50` points
- If you solve a puzzle within 48h, you get `30` points
- If you solve a puzzle anytime before the end of the contest, you get `20` points

The tie breaker would be the _median exact solve time for solved puzzles_.

**Comment by pil0u** [comment]
Following up on #83 and issues with the city leaderboard.

The median # of Alumni per city is `122`, and _all_ cities have more than `10`.
The median % of city participation (# of AoC participants divided by # of Alumni for each city) was `2.5%` in 2021.

**Proposition: we take the average of the total score of the `top 5%` of total number of city's Alumni (rounded up), with a minimum of `top 10` participants.**

Cologne has `40` Alumni, 5% = 2, so we average the score of their top `10` participants.
Paris has `2795` Alumni, 5% = 139.75, so we average the score of their top `140` participants.

Suppose Paris has only 100 participants. Then we consider the missing 40 participants with a score of `0` to compute the average.

**Comment by pil0u** [comment]
During 2021 event:
- `50%` of the completions happened within the first 6 hours upon puzzle release
- `66.7%` within 12 hours
- `77.1%` within 24 hours <- Bad news
- `82.7%` within 48 hours
- `87.2%` within 72 hours
- `93.4%` within 7 days

Also note that `56.9%` of the completions were during the first 7 days.

There's a bias because the competition was based on release time.
There's a bias for people who tried but didn't finish a puzzle.

I still want to definitely erase the timezone problems from the casual ranking, thus cannot consider less than 24 hours for most points attribution. Which means a lot of ties. But these people can enter the Ladder of Insanity for competition.
- `50` points if you solve a puzzle in less than 24 hours
- `50 - n_hours + 24` points if you solved it between 24 and 48 hours
- `25`points if you solve a puzzle in more than 48 hours

In other words:
`score = max(50 - max(n_hours - 24, 0), 25)

**Comment by pil0u** [comment]
## Tie-breakers

### Casual solo

- Score
- Number of challenges solved in less than 24 hours
- Number of challenges solved in less than 48 hours

### Squads

- Score
- Total number of challenges solved in less than 24 hours by squad members
- Total number of challenges solved in less than 48 hours by squad members

### Cities

- Score
- Total number of challenges solved in less than 24 hours by their top contributors (top 5% or top 10)
- Total number of challenges solved in less than 48 hours by their top contributors (top 5% or top 10)

### Ladder of Insanity

- Score
- Total time to solve challenges

**Comment by pil0u** [comment]
## Number of squad members

I think it's already very committing to have to assemble a team manually. I expect for a number of squads to be made of random people or people who don't know themselves very well, assembling. By definition, it will be easier for them to assemble a team of 4 than a team of 5.

Only the top 26 people were tied at number of stars last year. It means that even if they teamed up together, we would have 6 tied squads at top, which is acceptable.

I vote for `4`.
